javascript/ql/src/semmle/javascript/security/UselessUseOfCat.qll

```diff
@@ -0,0 +1,339 @@
+/**
+ * Provides predicates and classes for working with useless uses of the unix command `cat`.
+ */
+
+import javascript
+import Expressions.ExprHasNoEffect
+import Declarations.UnusedVariable
+
+/**
+ * A call that executes a system command.
+ * This class provide utility predicates for reasoning about command execution calls.
+ */
+private class CommandCall extends DataFlow::InvokeNode {
+  SystemCommandExecution command;
+
+  CommandCall() { this = command }
+
+  /**
+   * Holds if the call is synchronous (e.g. `execFileSync`).
+   */
+  predicate isSync() { command.isSync() }
+
+  /**
+   * Gets an argument to this command execution that specifies the argument list to the command.
+   */
+  DataFlow::Node getArgumentList() { result = command.getArgumentList() }
+
+  /**
+   * Gets the callback (if it exists) for an async `exec`-like call.
+   */
+  DataFlow::FunctionNode getCallback() {
+    not this.isSync() and result = getLastArgument().getALocalSource()
+  }
+
+  /**
+   * Holds if the executed command execution has an argument list as a separate argument.
+   */
+  predicate hasArgumentList() { exists(command.getArgumentList()) }
+
+  /**
+   * Gets the data-flow node (if it exists) for a options argument for an `exec`-like call.
+   */
+  DataFlow::Node getOptionsArg() {
+    exists(int n |
+      n >= 1 and
+      // if there is a command-list, then the options is at least the third argument.
```

Comment:
I agree that this probably holds in practice, but it seems safer to introduce `abstract DataFlow::Node getOptionsArg()` into SystemCommandExecution`.